### ISSUE TITLE
ODataException("The specified OData query has an unterminated string literal.", Text, TextPosition); line 244 if your query value is "Mike & Doe"

### ISSUE THREAD

**Issue by Vuminkosi** [Azure/azure-mobile-apps#685]
## Describe the bug

When is query the SqliteDataStore IOfflineTable with a query expression which one of the random values include & symbol i get this error message The specified OData query has an unterminated string literal.

## To Reproduce

// This is the expression query o n the DataService
var commodity = "Building & Fencing"; 
  var query =_offlineTable.CreateQuery().Where(x=>x.Group ==commodity ).OrderByDescending(x => x.UpdateDate);
   var models = await query.ToListAsync(cancellationToken: cancellationToken);
 

The specified OData query has an unterminated string literal. 

## Expected behavior
Return a list of Items, this is even if i terminate the string literal properly

## What platforms?   
  * .NET Runtime Environment (WPF, UWP, Xamarin.Forms, etc.): 
This happens on a Wpf App windows 10 im using Microsoft.Datasync.Client.SQLiteStore v5.1.5
## Additional context
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionLexer.CompleteStringLiteral()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionLexer.NextToken()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseComparison()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseLogicalAnd()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseLogicalOr()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseParenExpression()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParsePrimary()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseMultiplicative()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseAdditive()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseComparison()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseLogicalAnd()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseLogicalOr()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseParenExpression()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParsePrimary()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseMultiplicative()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseAdditive()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseComparison()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseLogicalAnd()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseLogicalOr()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseParenExpression()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParsePrimary()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseMultiplicative()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseAdditive()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseComparison()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseLogicalAnd()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseLogicalOr()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseParenExpression()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParsePrimary()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseMultiplicative()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseAdditive()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseComparison()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseLogicalAnd()
   at Microsoft.Datasync.Client.Query.OData.ODataExpressionParser.ParseLogicalOr()
   at Microsoft.Datasync.Client.Query.QueryDescription.Parse(String tableName, String query)
   at Microsoft.Datasync.Client.Offline.SyncContext.<GetNextPageAsync>d__34.MoveNext()
   at Microsoft.Datasync.Client.Table.OfflineTable`1.<GetNextPageAsync>d__33`1.MoveNext()
   at Microsoft.Datasync.Client.Table.FuncAsyncPageable`1.<AsPages>d__2.MoveNext()
   at Microsoft.Datasync.Client.Table.FuncAsyncPageable`1.<AsPages>d__2.System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult(Int16 token)
   at Microsoft.Datasync.Client.Table.AsyncPageable`1.<GetAsyncEnumerator>d__2.MoveNext()
   at Microsoft.Datasync.Client.Table.AsyncPageable`1.<GetAsyncEnumerator>d__2.MoveNext()
   at Microsoft.Datasync.Client.Table.AsyncPageable`1.<GetAsyncEnumerator>d__2.System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult(Int16 token)
   at Microsoft.Datasync.Client.Extensions.LinqExtensions.<ToZumoListAsync>d__6`1.MoveNext()
   at Microsoft.Datasync.Client.Extensions.LinqExtensions.<ToZumoListAsync>d__6`1.MoveNext()
   at System.Threading.Tasks.ValueTask`1.get_Result()
   at System.Runtime.CompilerServices.ValueTaskAwaiter`1.GetResult() 
=>DataService.cs:line 289

**Comment by Vuminkosi** [comment]
have been getting the exception for a week so i went through the repo i realize  that you are using a do-while loop to complete the string literal. This might cause a problem if the string literal contains a special character that is also used as a quote, such as ' or ". For example, if your query value is "Mike & Doe", then your loop will end at the first " and throw an exception. the only possible solution im seeing is to use a while loop instead and check for the escape character \ before advancing to the next occurrence of the quote

**Comment by adrianhall** [comment]
Please fill in the form that was provided.  It provides valuable information that enables me to diagnose the problem.

**Comment by Vuminkosi** [comment]
Thank you for the response, i have updated the form as requested

**Comment by Vuminkosi** [comment]
**Okay i got away with adding the following extension to terminate all string literals, works fine now**

```
 public static string TSL( this string stringValue)
        {
            if (string.IsNullOrWhiteSpace(stringValue))
                return stringValue; 

            return stringValue.Replace("&", "%26")
                  .Replace("+", "%2B")
                  .Replace("/", "%2F");
        }
```

**Comment by adrianhall** [comment]
I think this is actually a bug.  I'm going to re-open this, add tests to the set, and see what I can do to fix it.

**Comment by Vuminkosi** [comment]
Would greatly appreciate not using that extension

**Comment by adrianhall** [comment]
The problem was that the `ToODataString()` method (in Query/QueryDescriptionExtensions.cs) was not escaping the filter - complete oversight.  Once I figured out what was going on, it was an easy fix.  Running tests to validate the fix now.

I've also added a new KitchenSink test to ensure this case gets tested on a regular basis.

**Comment by adrianhall** [comment]
Released with v5.1.6 - make sure you remove your encoding extension method as you update the library.  If you don't, your string will be encoded twice, which will result in an error.

**Comment by Vuminkosi** [comment]
Thank you for the update
